blog/_src/posts/2018-12-02-java-transient.md

```diff
@@ -0,0 +1,361 @@
+ to the Java 1.4.0 type system.
+The benefit of generics is that a programmer can:
+ write one class definition,
+ use the definition in a few different contexts,
+ and receive specific feedback from the type checker in each context.
+
+### Review: generic types
+
+Suppose we want to write a `Box` class that holds some kind of value;
+ the value could be an `Integer` or a `String` or anything else.
+Here is a pre-generics definition:
+
+```
+class Box {
+  private Object val;
+
+  public Box(Object val) { this.set(val); }
+
+  public void set(Object val) { this.val = val; }
+
+  public Object get() { return this.val; }
+}
+```
+
+With this definition is it possible to make boxes that hold different types
+ of values:
+
+```
+// good!
+Box iBox = new Box(new Integer(4));
+Box sBox = new Box(new String("X"));
+```
+
+but it is also possible to "change the type" of the contents of a `Box`:
+
+```
+// maybe bad!
+iBox.set(new String("not a number"));
+```
+
+and some calls to `get` must be followed by a type cast:
+
+```
+// annoying!
+((String) sBox.get()).charAt(0);
+```
+
+**With generics**, we can give a name (e.g. `ValType`) to "the type of the value inside a box":
+
+```
+class GBox<ValType> {
+  private ValType val;
+
+  public GBox(ValType val) { this.set(val); }
+
+  public void set(ValType val) { this.val = val; }
+
+  public ValType get() { return this.val; }
+}
+```
+
+and now we can tell the type checker to check different boxes differently
+ (to opt-out, declare the type as `GBox<Object>`):
+
+```
+GBox<Integer> iBox = new GBox<Integer>(new Integer(0));
+GBox<String> sBox = new GBox<String>(new String("A"));
+
+// iBox.set(new String("not a number")); // Type Error, good!
+
+sBox.get().charAt(0); // no cast, good!
+```
+
+
+### Backwards compatibility & danger
+
+Java generics are backwards-compatible with older code.
+This means that pre-generics code can interact with instances of a generic
+ class (and vice-versa, generic code can interact with pre-generics classes).
+Since pre-generics code is not aware of type parameters, these interactions
+ are potentially unsafe.
+For example, a pre-generics method can change the type of a `GBox`:
+
+```
+// Java 1.4.0 method
+public static void evil(GBox b) { b.set(666); }
+
+// Java 1.5.0 method
+public static void test() {
+  GBox<String> sBox = new GBox<String>(new String("A"));
+  evil(sBox); // OK, but generates unchecked warning
+  sBox.get().charAt(0);
+}
+```
+
+The code above passes the type checker (with a warning about the `evil` method),
+ and so it _seems_ as though running the code will run the nonsense
+ method call `666.charAt(0)` and lead to evil behavior.
+The actual result, however, is a cast error immediately after the call
+ `sBox.get()` returns.
+
+Based on the cast error, we can tell that
+ the compiler does not trust the type `GBox<String>` and
+ inserts a run-time check that the result of a `.get()` matches the type
```

Comment:
The cast that Java inserts at these sites is a standard Java cast, which (due to the design of the object system) fully establishes the types of the referenced object (up to generics). It would make sense to me to mention this difference, especially in the discussion of the guarantees that Java is able to provide after such a dereference.